### PR TITLE
Packet Controlled Manufacturers

### PR DESCRIPTION
**Pull request by Conchuckter**
## About the PR 
Adds packet integration to manufacturers/fabricators, also seems to fix a couple problems with pausing the queue and having the time left go to like 500 seconds.

Manufacturers use the wired DataNet, however they are not connected by default, this is sorta intended so some nerd doesn't decide to spam print saxophones from across the station.

**Commands:**
- **Help:** Returns help index or information if index is specified.
- **Status:** Returns state of manufacturer.
- **Queue:** Returns the entire queue of the manufacturer separated by commas.
- **Add:** Adds a blueprint to the manufacturers queue.
- **Remove:** Removes a blueprint from the manufacturers queue.
- **Clear:** Clears the manufacturers queue.
- **Speed:** Sets the manufacturers production speed.
- **Resume:** Resumes manufacturing a paused blueprint.
- **Pause:** Pauses the current blueprint.
- **Repeat:** Turns on/off manufacturing repeat.

If a command requires an arg it can be accessed with `data` the only exception being help.

Errors and other messages are sent via `term_message` do note that this does not actually have terminal integration, and it was just used for consistency.

## Why's this needed? 
Automation could be interesting, especially with mechcomp inserters etc.

Expanding packet integration is way cool.

## Changelog 

```changelog
(u)Conchuckter
(*)Manufacturers are now packet integrated, become a nerd today!
```

### CODE REVIEW

**Comment by zjdtmkhzt** [comment]
The description says it fixes issues related to pausing, presumably that is the change that makes it so it uses the TIME macro instead of world.time?

**Comment by Conchuckter** [comment]
Yup, I've had problems where after pausing the queue `time_left` would be set to something like 500 seconds. From my testing this has fixed it but I'm not entirely sure.

Not sure if its actually been reported.

File: code/obj/machinery/manufacturer.dm [hunk 1]
```diff
@@ -1214,6 +1226,213 @@
...
+
+				post_signal(list("address_1" = sender, "sender" = src.net_id, "command" = "term_message", "data" = "ACK#CLEARED"))
+
+			if ("remove")
+				var/operation = text2num_safe(signal.data["data"])
+				if (!isnum(operation) || src.queue.len < 1 || operation > src.queue.len)
+					post_signal(list("address_1" = sender, "sender" = src.net_id, "command" = "term_message", "data" = "ERR#BADOPERATION"))
+					return
+
+				if(world.time < last_queue_op + 5)
```
Comment (TobleroneSwordfish) [inline code/obj/machinery/manufacturer.dm:1364]: Consider using the `TIME` define here, to prevent issues with lag slowing things down.

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
one step closer to true factorio SS13

File: code/obj/machinery/manufacturer.dm [hunk 2]
```diff
@@ -3,6 +3,8 @@
 #define WIRE_POWER 2
 #define WIRE_MALF 3
 #define WIRE_SHOCK 4
+#define MAX_SPEED
+#define MAX_SPEED_HACKED
```
[the remark this answers is not in the document]
Comment (TobleroneSwordfish) [reply code/obj/machinery/manufacturer.dm:7]: should also `#undef` these at the bottom of the file

File: code/obj/machinery/manufacturer.dm [hunk 3]
```diff
@@ -1214,6 +1235,219 @@
...
+				else
+					post_signal(list("address_1" = sender, "sender" = src.net_id, "command" = "term_message", "data" = "ERR#WORKING"))
+
+			if ("remove")
+				var/operation = text2num_safe(signal.data["data"])
+				if (!isnum(operation) || length(src.queue) < 1 || operation > length(src.queue))
+					post_signal(list("address_1" = sender, "sender" = src.net_id, "command" = "term_message", "data" = "ERR#BADOPERATION"))
+					return
+
+				if(world.time < last_queue_op + 5)
```
Comment (pali6) [inline code/obj/machinery/manufacturer.dm:1379]:
```suggestion
				if(TIME < last_queue_op + 5)
```

File: code/obj/machinery/manufacturer.dm [hunk 4]
```diff
@@ -1214,6 +1235,219 @@
...
+					return
+
+				else
+					last_queue_op = world.time
```
Comment (pali6) [inline code/obj/machinery/manufacturer.dm:1383]:
```suggestion
					last_queue_op = TIME
```

**Review by pali6 — APPROVED** [review]

**Comment by pali6** [comment]
Looks good to me, waiting for @MarkNais's review.

**Review by MarkNais — CHANGES_REQUESTED** [review]
One minor-issue: because the `help` command responds with `command=help` it causes a temporary feedback loop. See screenshot below. It thankfully stops (somehow) without being noticeable on local-testing, but still not something I want on production.
So I've suggested all of the responses being `command=term_message` .

image

File: code/obj/machinery/manufacturer.dm [hunk 5]
```diff
@@ -1214,6 +1235,218 @@
...
+
+				post_signal(help)
+				return
+
+			if ("status")
+				post_signal(list("address_1" = sender, "sender" = src.net_id, "command" = "status", "data" = "mode=[src.mode]&speed=[src.speed]&timeleft=[src.time_left]&repeat=[src.repeat]"))
+				return
+
+			if ("queue")
+				post_signal(list("address_1" = sender, "sender" = src.net_id, "command" = "queue", "data" = src.queue.Join(",")))
```
[the remark this answers is not in the document]
Comment (Conchuckter) [reply code/obj/machinery/manufacturer.dm]: Just removed the command field from these packets entirely, as it is more consistent with other devices.

**Pull request merged**